prow/scheduler/strategy/default_test.go

```diff
@@ -59,3 +62,42 @@ func TestPassthrough(t *testing.T) {
 		})
 	}
 }
+
+func TestGet(t *testing.T) {
+	for _, tc := range []struct {
+		name          string
+		config        config.Scheduler
+		knownClusters map[string]rest.Config
+		wantStrategy  strategy.Interface
+	}{
+		{
+			name:         "Default to passthrough",
+			wantStrategy: &strategy.Passthrough{},
+		},
+		{
+			name:         "Get round robin",
+			config:       config.Scheduler{RoundRobin: &config.RoundRobinScheduling{}},
+			wantStrategy: &strategy.RoundRobin{},
+		},
+		{
+			name:         "Get failover",
+			config:       config.Scheduler{Failover: &config.FailoverScheduling{}},
+			wantStrategy: &strategy.Failover{},
+		},
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			cfg := &config.Config{ProwConfig: config.ProwConfig{Scheduler: tc.config}}
+			strategy := strategy.Get(cfg, tc.knownClusters)
+
+			strategyType := reflect.TypeOf(strategy)
```

Comment:
Let's not compare the type of the struct we will get and instead compare the exact struct with the expected one.

Reply:
```golang
if !reflect.DeepEqual(tc.wantStrategy, strategy) {
    t.Errorf("Want %+v but got %+v", tc.wantStrategy, strategy)
}
```
Are you ok with this one? `Get` returns an interface, I'm not sure whether we can compare the two actual struct without using the reflection.

Reply:
You can compare exactly the structs, if you will define the expected variable as the interface. We also need to test cases of different configuration of the same strategies as well.

Reply:
There are some edge cases that makes structs incomparable:
- `RoundRobin` holds a slices, and slices are not comparable using `==`
- Interfaces holding pointer are equal only if they points to the same struct, so this will always fail:
```golang
var a strategy.Interface = &strategy.RoundRobin{}
var b strategy.Interface = &strategy.RoundRobin{}
if a != b {
  // Fail
}
```

We could use:
```golang
cmd.Diff(a, b, cmpopts.IgnoreUnexported(strategy.RoundRobin{}))
```
but there are some private fields we have to ignore or it is going fail.

Reply:
Yea you can ignore the unexported fields, but I don't think there is another case that makes the structs incomparable.  The test case should compare the exact result that we expect. The slice inside the `RoundRobin` shouldn't be a problem to compare.

Reply:
A unit test that attempts to compare/assert over unexported fields is indicative of a fragile factoring - why should the caller of this function have opinions on unexported fields?

The method returns an interface - the caller (and tests) are not valid to have opinions on how that interface is fulfilled. Let's test this functionally - given a config, get a scheduler, and ask it to schedule. Ensure that given the right config, the right behavior comes out the other end. Implementation details like the type under the interface are not a super useful thing to test.